x-pack/plugins/enterprise_search/public/applications/app_search/components/engine/engine_router.tsx

```diff
@@ -69,31 +69,30 @@ export const EngineRouter: React.FC = () => {
     },
   } = useValues(AppLogic);
 
+  const { engineName: engineNameFromUrl } = useParams() as { engineName: string };
   const { engineName, dataLoading, engineNotFound } = useValues(EngineLogic);
   const { setEngineName, initializeEngine, clearEngine } = useActions(EngineLogic);
 
-  const { engineName: engineNameParam } = useParams() as { engineName: string };
-  const engineBreadcrumb = [ENGINES_TITLE, engineNameParam];
-
-  const isEngineInStateStale = () => engineName !== engineNameParam;
-
   useEffect(() => {
-    setEngineName(engineNameParam);
+    setEngineName(engineNameFromUrl);
     initializeEngine();
     return clearEngine;
-  }, [engineNameParam]);
+  }, [engineNameFromUrl]);
 
   if (engineNotFound) {
     setQueuedErrorMessage(
       i18n.translate('xpack.enterpriseSearch.appSearch.engine.notFound', {
-        defaultMessage: "No engine with name '{engineNameParam}' could be found.",
-        values: { engineNameParam },
+        defaultMessage: "No engine with name '{engineName}' could be found.",
+        values: { engineName },
       })
     );
     return <Redirect to={ENGINES_PATH} />;
   }
 
-  if (isEngineInStateStale() || dataLoading) return <Loading />;
+  const isLoadingNewEngine = engineName !== engineNameFromUrl;
+  if (isLoadingNewEngine || dataLoading) return <Loading />;
+
+  const engineBreadcrumb = [ENGINES_TITLE, engineName];
```

Comment:
Similarly, I wanted to simplify `engineBreadcrumb` to just use `engineName` (and also to think of engineName as a source of truth except initializeEngine), so I moved this to after the Loading component to reduce confusion/keep the logic together.